File: Algorithms/494_TargetSum.py
class Solution(object):
    def findTargetSumWays(self, nums, S):
        """
        :type nums: List[int]
        :type S: int
        :rtype: int
        """
        m = {}
        return self.findTargetSumWaysInterval(nums, 0, len(nums)-1, S, m)
    
    def findTargetSumWaysInterval(self, nums, start, end, S, m):
        if start > end:
            return S == 0
        if (start, S) in m:
            return m[(start, S)]
        t = self.findTargetSumWaysInterval(nums, start+1, end, S+nums[start], m) + self.findTargetSumWaysInterval(nums, start+1, end, S-nums[start], m)
        m[(start, S)] = t
        return t

File: Algorithms/java/494.TargetSum.java
class Solution {
    public int findTargetSumWays(int[] nums, int S) {
        return findTargetSumWays(nums, 0, nums.length-1, S);
    }
    private int findTargetSumWays(int[] nums, int start, int end, int S) {
        if (start > end) {
            if (S == 0) return 1;
            else return 0;
        }
        return findTargetSumWays(nums, start+1, end, S+nums[start]) + findTargetSumWays(nums, start+1, end, S-nums[start]);
    }
}
